### PR TITLE
NO-ISSUE: dev-guide/release-blocker-definition: Distinguish X.Y.0 GA from X.Y.Z patch-fixes

### PR DESCRIPTION
**Pull request by wking**
Following up on [this thread][1], point out that the criteria for delaying patch releases are much higher, because we don't want to stand between users and any _other_ bugfixes that are ready to ship.  Also give folks some pointers at mitigation options that are less severe than "withhold the patch fix". In some cases, there are internal process docs (e.g. for KCS creation), but I dunno if I should be linking those in this public doc?

[1]: https://github.com/openshift/enhancements/pull/1926#discussion_r2710169713

### CODE REVIEW

File: dev-guide/release-blocker-definition.md [hunk 1]
```diff
@@ -72,6 +74,33 @@ We know there will be edge cases.  When those come up:
...
+There are multiple options for protecting users when a release ships with a known issue.
+They include:
+
+* Known issues in the release notes, e.g. [known issues with 4.20][4.20-known-issues].
+* [Knowledgebase solutions][kcs].
+* [Insights rules][4.20-insights] for the subset of the fleet that enables remote-health reporting.
+* [Conditional update risk declarations][4.20-conditional-updates], see [the assessment process](/enhancements/update/update-blocker-lifecycle/README.md) and [the distribution tooling](/enhancements/update/targeted-update-edge-blocking.md).
+
+Both Insights rules and conditional update risk declarations can sometimes be scoped down to the exposed subset of the fleet, to reduce distracting the subset of the fleet that is not exposed.
+Both Insights rules and condtional update risk declarations are specific to updating into the risk, and neither provides pre-install warning for users consdering installing an exposed release.
```
Comment (dgoodwin) [inline dev-guide/release-blocker-definition.md]:
```suggestion
Both Insights rules and conditional update risk declarations are specific to updating into the risk. There is no warning prior to a fresh install of an exposed release.
```

File: dev-guide/release-blocker-definition.md [hunk 2]
```diff
@@ -72,6 +74,33 @@ We know there will be edge cases.  When those come up:
   * Service Delivery Architecture Overview  
   * Stand Up
 
+### When to set Release Blocker to `Approved` in an X.Y.Z OCPBUG
+
+When preparing a new X.Y.Z patch release shipping bug fixes, our ability to delay the release is much more constrained.
+If a bug is so terrible that _no cluster_ could possibly benefit from the impacted nightlies, setting `Release Blocker` to `Approved` is appropriate.
+But if there are even small subsets of clusters that would not be exposed to the OCPBUG, or if there are steps a cluster-admin could take to mitigate exposure, we prefer releasing the patch release anyway.
+This avoids delaying access to other bugfixes in the the clusters that could survive with the OCPBUG being considered.
```
Comment (dgoodwin) [inline dev-guide/release-blocker-definition.md:82]: In this section I presume we're talking about NEW regressions in the .z release we've discovered, and not some bug that's already in the wild correct?

Already in the wild, I would think rarely would we have to hold a .z for an approved release blocker.

If we discover something has gone wrong with a fix in the .z, and it will break functionality for a subset of customers in a new way, I worry that the mitigations below convey a lack of confidence in our own product and systems that we then make very apparent to customers and expect them to accommodate on every update.

For one I worry they won't read or do sufficient due diligence to determine if the problem will affect them. However even if we could induce that behaviour given how rare this situation is likely to arise, we're still expecting them to see an update and not have confidence it's safe to apply unless THEY take some action to find out if it's going to break them. When those updates go out, we should have confidence we're not making things worse. 

I'm new to this and happy to discuss more, and I would hope this has been a VERY rare thing historically, but initially I think we should not regress functionality in .z's.

File: dev-guide/release-blocker-definition.md [hunk 3]
```diff
@@ -72,6 +74,33 @@ We know there will be edge cases.  When those come up:
...
+
+In some cases, a hot fix is about to merge, or is percolating through nightlies, or is otherwise expected to show up very soon in the release pipeline.
+Setting `Release Blocker` to `Approved` in those cases is acceptable as a way to attract ART attention and ask them to wait for that fix.
+They may wait for the fix, or they may decide to build a release anyway, even without the fix, depending on capacity and how much time remains for final QE testing before the planned Errata-publication time.
```
Comment (dgoodwin) [inline dev-guide/release-blocker-definition.md]: If ART can ignore release blocker, we have no mechanism I can see to actually stop a .z from going out. I would propose release blocker approved means no release going out, and we work out the details to decide if it's safe to remove that flag. (which will soon require approvals hopefully)

Comment (wking) [reply dev-guide/release-blocker-definition.md]: I was trying to leave room for [Scott's][1]:

>  TBH this lever hasn't actually been particularly successful in getting ART to delay promotion...

But I'm ok tightening it up to reflect aspirations instead of current reality, and then working post-merge to try and catch reality up with the aspirations.  Is that what you're looking for?

[1]: https://github.com/openshift/enhancements/pull/1926#discussion_r2713727599

File: dev-guide/release-blocker-definition.md (hunk 2 above)
Comment (wking) [reply dev-guide/release-blocker-definition.md:82]: > In this section I presume we're talking about NEW regressions in the .z release we've discovered, and not some bug that's already in the wild correct?

I was talking about bugs that were known to affect the current X.Y branch tip, without distinguishing based on whether existing X.Y.Z were exposed.  More later in this comment about why I don't think we want to make a distinction there.

> ...I worry that the mitigations below convey a lack of confidence in our own product and systems that we then make very apparent to customers...

One response to the reputational risk is that, if the rate of regressions within a z stream rises to embarrassing levels, we invest more effort into CI/QE to get the rate back down to something we can stomach.  I'm not a fan of ignoring known issues that have already shipped because we're feeling embarrassed; I think we need to own up to those and let users know before more get bit by the known issue.  I accept that sometimes our ability to scope a risk declaration to the subset of exposed clusters is small, and we may decide to use other mechanisms to mitigate, depending on how we want to balance the risk of a cluster-admin who is exposed missing the warning against the distraction of a cluster-admin who is not exposed having the irrelevant-for-them warning waived around.

And part of the pitch I'm making is that there are going to be _other_ bugs where we have verified fixes.  And getting those fixes out faster to the subset of the fleet that can safely benefit from them seems like it would help our reputation for security and stability.  And the fact that there are other fixes in the pipe that will benefit the not-exposed-to-the-regression portion of the fleet is why I don't think we want to distinguish between "already wild" and "just regressed" in our z-stream-release-blocker policy.

> ...expect them to accommodate on every update.

At least _update_ mitigations include Insights and conditional update risks, which can surface issues to the affected subset of cluster-admins, without requiring them to manually reach out and hunt for known-issues.

> ...I would hope this has been a VERY rare thing historically, but initially I think we should not regress functionality in .z's.

We're usually pretty good for things in the main CI/QE wheelhouse.  But the whole cluster-config space is combinatorically large, and exhaustive test coverage is impossible, and things slip through the cracks (mostly for corner-case configs) fairly often.  See [here][1] for me doing an audit of our [public update risk declarations][2] from back in the 4.13/4.14 timeframe.

[1]: https://github.com/openshift/enhancements/pull/1559/files#diff-0734cd0d197b626040a42b9e3b51d5c2fa59fb5f6d5d4b96851008cb1db08378R358-R392
[2]: https://github.com/openshift/cincinnati-graph-data?tab=readme-ov-file#block-edges

File: dev-guide/release-blocker-definition.md (hunk 1 above)
Comment (wking) [reply dev-guide/release-blocker-definition.md]: I pushed the "condtional" -> "conditional" fix with f8e36ad8 -> 030af07d ; thanks!

> There is no warning prior to a fresh install of an exposed release.

I didn't pick up this part.  What I'm trying to say in this sentence is that these two options (Insights and conditional updates) aren't relevant for install.  I'm not ruling out _other_ mitigations that might be helpful for install (e.g. known issues and KCS could help for install, although they would require a fair bit of motivation to work for _pre-install_).  Is that trying to slice it too fine?  Or maybe it can work, but I should add a preceding sentence that points out that some mechanisms are better fits for some situations, to get folks in that headspace before I start getting into per-mechanism limitations?  Or spread out with subsections for each mitigation option, instead of trying to cover them all in a list and sentence-or-two?  Or...?

Comment (wking) [reply dev-guide/release-blocker-definition.md]: alternatively, maybe I should give up on each sentence being able to stand alone, and say something like:

> Both Insights rules and conditional update risk declarations can sometimes be scoped down to the exposed subset of the fleet, to reduce distracting the subset of the fleet that is not exposed.  They are specific to updating into the risk.  Neither provides pre-install or post-failed-install warning for users considering installing an exposed release.

?

File: dev-guide/release-blocker-definition.md (hunk 3 above)
Comment (mffiedler) [reply dev-guide/release-blocker-definition.md]: +1 to Devan's suggestion that Release Blocker Approved have enough teeth to stop a release until discussion has occurred.

File: dev-guide/release-blocker-definition.md (hunk 2 above)
Comment (mffiedler) [reply dev-guide/release-blocker-definition.md:82]: We do consider "already in the wild" status to make a bug an unlikely blocker candidate for x.y.0.  The thinking is that the bug will be dealt with in the  z-stream workflow (fixed or left un-fixed depending on severity).  I agree that we  shouldn't disqualify x.y.z bugs as potential blockers if they are already in the wild.   Something like the OLM catalog issues is an extreme example of something in the wild that had to be addressed as a blocker.  It should not be common to make an already-released bug a blocker, but it can be a tool to use if needed.

File: dev-guide/release-blocker-definition.md (hunk 3 above)
Comment (JoelSpeed) [reply dev-guide/release-blocker-definition.md]: Coming back to this since Trevor highlighted it to me yesterday

If setting release blocker to approved doesn't block the release, we have a fundamental issue with the process that we should fix. I think that's a separate discussion to this PR though.

This paragraph in its own right strikes me as odd. It's not deterministic about whether a bug is a blocker or not. "You may want to do this to try and squeeze your bug in" doesn't feel like something I'd want to promote the org to be doing - gaming the system effectively. The bug is either a blocker, or it is not, and if it's not, and you miss, there will be another build soon

Comment (wking) [reply dev-guide/release-blocker-definition.md]: > If setting release blocker to approved doesn't block the release, we have a fundamental issue with the process that we should fix...

In a world where ART _does_ have tooling that reliably blocks an X.Y.Z from releasing when there's an open `Release Blocker: Approved` against that X.Y.Z, where is the block?  Does ART check for release-blockers before requesting a new X.Y.Z release image get built?  After it's built, but before they request the signature(s)?  After it's signed, but before they push the errata?  All of those?  Other?  It seems sticky.

>  "You may want to do this to try and squeeze your bug in" doesn't feel like something I'd want to promote the org to be doing...

Folks do it today.  I'm personally fine having the docs here go whichever way folks like, and I expect folks sitting on an open pull for a hot bug will continue to do both "just merge as fast as I safely can" and "claim `Release Blocker: Approved` to try to slow the train down a bit to buy time for my merge", regardless of what we doc here, and regardless of whether ART is actually watching guardrails based on `Release Blocker: Approved` queries or not.

Comment (wking) [reply dev-guide/release-blocker-definition.md]: > I'm personally fine having the docs here go whichever way folks like...

And to rephrase this, please suggest the text you want to see, and I'll probably be fine just dropping it in instead of what I currently have ;)

Comment (JoelSpeed) [inline dev-guide/release-blocker-definition.md]: IMO I'd rather we guide folks away from the idea that they can use Release blocker to try and catch a release that's about to be cut. It doesn't scale well when the whole org uses this, and is probably the reason ART has to ignore the signal in the first place.

```suggestion
The `Release Blocker` is not a tool to be used slow down releases just because a hot fix is _about_ to merge. Using `Release Blocker` in this way does not scale well to the org level and results in `Release Blocker` being an unreliable signal (ART must ignore it).

The determination of `Release Blocker` should be solely based on the impact to end users, and not based on trying to time a fix into a particular release.
```

Comment (wking) [reply dev-guide/release-blocker-definition.md]: I've pushed 030af07d -> a30e8e3c to rebase and set a version of this^.

```diff
diff --git a/dev-guide/release-blocker-definition.md b/dev-guide/release-blocker-definition.md
index e1af8623..a64a1eec 100644
--- a/dev-guide/release-blocker-definition.md
+++ b/dev-guide/release-blocker-definition.md
@@ -4,7 +4,7 @@ authors:
   - "@mffiedler"
   - "@dgoodwin"
 creation-date: 2026-01-20
-last-updated: 2026-01-23
+last-updated: 2026-07-02
 status: informational
 ---
 
@@ -77,13 +77,13 @@ We know there will be edge cases.  When those come up:
 ### When to set Release Blocker to `Approved` in an X.Y.Z OCPBUG
 
 When preparing a new X.Y.Z patch release shipping bug fixes, our ability to delay the release is much more constrained.
-If a bug is so terrible that _no cluster_ could possibly benefit from the impacted nightlies, setting `Release Blocker` to `Approved` is appropriate.
+If a bug is so terrible that _no cluster_ could possibly benefit from the impacted nightlies, setting `Release Blocker` to `Approved` is appropriate (and this is the only situation when setting `Approved` for an X.Y.Z bug is appropriate).
 But if there are even small subsets of clusters that would not be exposed to the OCPBUG, or if there are steps a cluster-admin could take to mitigate exposure, we prefer releasing the patch release anyway.
 This avoids delaying access to other bugfixes in the the clusters that could survive with the OCPBUG being considered.
 
-In some cases, a hot fix is about to merge, or is percolating through nightlies, or is otherwise expected to show up very soon in the release pipeline.
-Setting `Release Blocker` to `Approved` in those cases is acceptable as a way to attract ART attention and ask them to wait for that fix.
-They may wait for the fix, or they may decide to build a release anyway, even without the fix, depending on capacity and how much time remains for final QE testing before the planned Errata-publication time.
+`Release Blocker: Approved` is not a tool to be used slow down releases just because a hot fix is _about_ to merge.
+Using `Release Blocker` in this way does not scale well, because at the OpenShift level, there are few windows with no hot fixes in flight.
+If a fix misses the current patch release, another patch release will be built soon, and the fix can go out on that later release.
 
 #### Mitigating exposure
 
```

Feel free to push back on my wording if you feel like I'm missing something that you'd covered in yours, or whatever.

**Comment by JoelSpeed** [comment]
This LGTM now, @dgoodwin can you check the most recent updates?

**Comment by dgoodwin** [comment]
/lgtm

Works for me.
